bundles/org.openhab.binding.senechome/src/main/java/org/openhab/binding/senechome/internal/SenecHomeHandler.java

```diff
@@ -211,50 +211,54 @@ private void refresh() {
-                    Units.KILOWATT_HOUR, DIVISOR_MILLI_TO_KILO);
-            updateQtyState(CHANNEL_SENEC_DISCHARGED_ENERGY_PACK4, response.battery.dischargedEnergy[3], 2,
-                    Units.KILOWATT_HOUR, DIVISOR_MILLI_TO_KILO);
-            updateDecimalState(CHANNEL_SENEC_CYCLES_PACK1, response.battery.cycles[0]);
-            updateDecimalState(CHANNEL_SENEC_CYCLES_PACK2, response.battery.cycles[1]);
-            updateDecimalState(CHANNEL_SENEC_CYCLES_PACK3, response.battery.cycles[2]);
-            updateDecimalState(CHANNEL_SENEC_CYCLES_PACK4, response.battery.cycles[3]);
-            updateQtyState(CHANNEL_SENEC_CURRENT_PACK1, response.battery.current[0], 2, Units.AMPERE);
-            updateQtyState(CHANNEL_SENEC_CURRENT_PACK2, response.battery.current[1], 2, Units.AMPERE);
-            updateQtyState(CHANNEL_SENEC_CURRENT_PACK3, response.battery.current[2], 2, Units.AMPERE);
-            updateQtyState(CHANNEL_SENEC_CURRENT_PACK4, response.battery.current[3], 2, Units.AMPERE);
-            updateQtyState(CHANNEL_SENEC_VOLTAGE_PACK1, response.battery.voltage[0], 2, Units.VOLT);
-            updateQtyState(CHANNEL_SENEC_VOLTAGE_PACK2, response.battery.voltage[1], 2, Units.VOLT);
-            updateQtyState(CHANNEL_SENEC_VOLTAGE_PACK3, response.battery.voltage[2], 2, Units.VOLT);
-            updateQtyState(CHANNEL_SENEC_VOLTAGE_PACK4, response.battery.voltage[3], 2, Units.VOLT);
-            updateQtyState(CHANNEL_SENEC_MAX_CELL_VOLTAGE_PACK1, response.battery.maxCellVoltage[0], 3, Units.VOLT,
-                    DIVISOR_MILLI_TO_ISO);
-            updateQtyState(CHANNEL_SENEC_MAX_CELL_VOLTAGE_PACK2, response.battery.maxCellVoltage[1], 3, Units.VOLT,
-                    DIVISOR_MILLI_TO_ISO);
-            updateQtyState(CHANNEL_SENEC_MAX_CELL_VOLTAGE_PACK3, response.battery.maxCellVoltage[2], 3, Units.VOLT,
-                    DIVISOR_MILLI_TO_ISO);
-            updateQtyState(CHANNEL_SENEC_MAX_CELL_VOLTAGE_PACK4, response.battery.maxCellVoltage[3], 3, Units.VOLT,
-                    DIVISOR_MILLI_TO_ISO);
-            updateQtyState(CHANNEL_SENEC_MIN_CELL_VOLTAGE_PACK1, response.battery.minCellVoltage[0], 3, Units.VOLT,
-                    DIVISOR_MILLI_TO_ISO);
-            updateQtyState(CHANNEL_SENEC_MIN_CELL_VOLTAGE_PACK2, response.battery.minCellVoltage[1], 3, Units.VOLT,
-                    DIVISOR_MILLI_TO_ISO);
-            updateQtyState(CHANNEL_SENEC_MIN_CELL_VOLTAGE_PACK3, response.battery.minCellVoltage[2], 3, Units.VOLT,
-                    DIVISOR_MILLI_TO_ISO);
-            updateQtyState(CHANNEL_SENEC_MIN_CELL_VOLTAGE_PACK4, response.battery.minCellVoltage[3], 3, Units.VOLT,
-                    DIVISOR_MILLI_TO_ISO);
+            // We are checking only one field here. It is assumed that other fields, like dischargedEnergy, cycles, etc.
+            // are available then, too, since these are typical for newer Senec systems with LiPo batteries.
+            if (response.battery.chargedEnergy != null) {
```

Comment:
`response.battery` could also be null. Furthermore, some channels requiring information not coming from `response.battery.chargedEnergy` are now skipped. Perhaps this block could be extracted into a private method with `response.battery` provided as parameter, and channels could be updated according to information available? This would avoid assumptions and also reduce nesting.

Reply:
Good point. I went for the path with the lesser effort first.
I would put some more time into this if I could test it myself.
Some strange error messages prevent me from updating my bindings in general. I've updated to 3.3M2, but I can't update my bindings from 3.1. Any idea?
-> https://community.openhab.org/t/openhab-3-3-milestone-discussion/132715/65?u=thopiekar

Reply:
Does it mean that your installation is broken after upgrading to 3.3M2? Unfortunately I don't know what could be the reason for this.

Reply:
I think so, but the message doesn't give me a clue how to fix it :disappointed:

Reply:
@thopiekar - just to be clear, intention was something like this:

```
if (response.battery != null) {
    updateBatteryStates(response.battery);
}
...

private void updateBatteryStates(SenecHomeBattery battery) {
    if (battery.chargedEnergy != null) {
        updateQtyState(CHANNEL_SENEC_CHARGED_ENERGY_PACK1, response.battery.chargedEnergy[0], 2, Units.KILOWATT_HOUR, DIVISOR_MILLI_TO_KILO);
        ....
    }
    if (battery.dischargedEnergy != null) {
        updateQtyState(CHANNEL_SENEC_DISCHARGED_ENERGY_PACK1, response.battery.dischargedEnergy[0], 2,
                        Units.KILOWATT_HOUR, DIVISOR_MILLI_TO_KILO);
        ...
    }
}
```

Remember also to check array sizes before indexing.

Reply:
But response.battery will never get null. As I said, the "skeleton" of the data is described in json/SenecHomeResponse.java as classes.
Even if my device doesn't send any battery info in its response, there will be empty children set by the class structure.
Therefore this might be an option:
```
    updateBatteryStates(response.battery);
...

private void updateBatteryStates(SenecHomeBattery battery) {
    if (battery.chargedEnergy != null) {
        updateQtyState(CHANNEL_SENEC_CHARGED_ENERGY_PACK1, battery.chargedEnergy[0], 2, Units.KILOWATT_HOUR, DIVISOR_MILLI_TO_KILO);
        ....
    }
    if (battery.dischargedEnergy != null) {
        updateQtyState(CHANNEL_SENEC_DISCHARGED_ENERGY_PACK1, battery.dischargedEnergy[0], 2, Units.KILOWATT_HOUR, DIVISOR_MILLI_TO_KILO);
        ...
    }
}
```

But this would be only a cosmetic thing in my point of view.

Reply:
OK. The exception from the issue mentioned is addressed by this PR, so resolving comment.